File: claude_exp/xv3/shared/stats.py
import threading
import time
from collections import deque
from datetime import datetime

WINDOWS = (30, 60, 180, 1800)
_MAX_WINDOW = max(WINDOWS)


class Stats:
    """Thread-safe rolling counters over windows [30, 60, 180, 1800] seconds."""

    def __init__(self, yellow_threshold_seconds=None):
        self._lock = threading.Lock()
        self._yellow_threshold_seconds = yellow_threshold_seconds
        self._sent_total = 0
        self._recv_total = 0
        self._sent_times = deque()
        self._recv_times = deque()
        self._last_recv_monotonic = None
        self._last_recv_datetime = None
        self._connections = {}
        self._gauges = {}

    def set_connection(self, name: str, connected: bool) -> None:
        with self._lock:
            self._connections[name] = connected

    def set_gauge(self, name: str, value) -> None:
        with self._lock:
            self._gauges[name] = value

    def record_sent(self) -> None:
        now = time.monotonic()
        with self._lock:
            self._sent_total += 1
            self._sent_times.append(now)
            self._trim(self._sent_times, now)

    def record_recv(self) -> None:
        now = time.monotonic()
        with self._lock:
            self._recv_total += 1
            self._recv_times.append(now)
            self._trim(self._recv_times, now)
            self._last_recv_monotonic = now
            self._last_recv_datetime = datetime.now()

    @staticmethod
    def _trim(times: deque, now: float) -> None:
        cutoff = now - _MAX_WINDOW
        while times and times[0] < cutoff:
            times.popleft()

    @staticmethod
    def _count_within(times: deque, now: float, window: int) -> int:
        cutoff = now - window
        count = 0
        for t in reversed(times):
            if t < cutoff:
                break
            count += 1
        return count

    def snapshot(self) -> dict:
        now = time.monotonic()
        with self._lock:
            self._trim(self._sent_times, now)
            self._trim(self._recv_times, now)

            result = {
                "sent_total": self._sent_total,
                "recv_total": self._recv_total,
            }
            for window in WINDOWS:
                result[f"sent_{window}s"] = self._count_within(self._sent_times, now, window)
                result[f"recv_{window}s"] = self._count_within(self._recv_times, now, window)

            if self._last_recv_monotonic is not None:
                result["seconds_since_last_recv"] = now - self._last_recv_monotonic
                result["last_recv_datetime"] = self._last_recv_datetime.strftime("%H:%M:%S")
            else:
                result["seconds_since_last_recv"] = None
                result["last_recv_datetime"] = None

            if self._yellow_threshold_seconds is not None:
                result["yellow_threshold_seconds"] = self._yellow_threshold_seconds

            if self._connections:
                result["connections"] = dict(self._connections)

            if self._gauges:
                result["gauges"] = dict(self._gauges)

        return result

File: claude_exp/routers/router_java/src/main/java/com/router/shared/Stats.java
package com.router.shared;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Thread-safe rolling counters over windows [30, 60, 180, 1800] seconds. Port of
 * router_py's shared/stats.py.
 */
public final class Stats {

    private static final int[] WINDOWS = {30, 60, 180, 1800};
    private static final DateTimeFormatter TIME_FMT = DateTimeFormatter.ofPattern("HH:mm:ss")
            .withZone(ZoneId.systemDefault());
    // Bounded by count, not time - a percentile only needs *some* recent sample of values, and a
    // fixed-size deque keeps memory and the snapshot()-time sort cost bounded even at high TPS.
    // Same "keep last N" convention as LogBuffer. Matches router_py's shared/stats.py.
    private static final int LATENCY_MAXLEN = 2000;

    private final Object lock = new Object();
    private final Integer yellowThresholdSeconds;
    private long sentTotal = 0;
    private long recvTotal = 0;
    private final Deque<Long> sentTimesMillis = new ArrayDeque<>();
    private final Deque<Long> recvTimesMillis = new ArrayDeque<>();
    private Long lastRecvMillis = null;
    private final Map<String, Boolean> connections = new LinkedHashMap<>();
    private final Map<String, Object> gauges = new LinkedHashMap<>();
    private final Map<String, Deque<Double>> latencies = new LinkedHashMap<>();

    public Stats(Integer yellowThresholdSeconds) {
        this.yellowThresholdSeconds = yellowThresholdSeconds;
    }

    public void setConnection(String name, boolean connected) {
        synchronized (lock) {
            connections.put(name, connected);
        }
    }

    public void setGauge(String name, Object value) {
        synchronized (lock) {
            gauges.put(name, value);
        }
    }

    public void recordSent() {
        synchronized (lock) {
            sentTotal++;
            sentTimesMillis.addLast(System.currentTimeMillis());
            trim(sentTimesMillis);
        }
    }

    public void recordRecv() {
        synchronized (lock) {
            long now = System.currentTimeMillis();
            recvTotal++;
            recvTimesMillis.addLast(now);
            trim(recvTimesMillis);
            lastRecvMillis = now;
        }
    }

    /**
     * Records one timing sample under a named bucket (e.g. "queue_wait", "crypto_rtt",
     * "downstream_rtt", "total" - see Dispatcher) for live per-hop latency visibility in
     * /stats, without waiting on an offline soak-test CSV. Available on every actor type
     * (Stats is shared), not just the router - any actor with a hop worth timing can call
     * this the same way.
     */
    public void recordLatency(String name, double valueMs) {
        synchronized (lock) {
            Deque<Double> bucket = latencies.computeIfAbsent(name, k -> new ArrayDeque<>());
            bucket.addLast(valueMs);
            while (bucket.size() > LATENCY_MAXLEN) {
                bucket.pollFirst();
            }
        }
    }

    /** Linear-interpolation percentile - matches router_py's _percentile helper. */
    private static double percentile(List<Double> sortedValues, double pct) {
        double k = (sortedValues.size() - 1) * pct;
        int f = (int) k;
        int c = Math.min(f + 1, sortedValues.size() - 1);
        if (f == c) {
            return sortedValues.get(f);
        }
        return sortedValues.get(f) + (sortedValues.get(c) - sortedValues.get(f)) * (k - f);
    }

    private static double round3(double v) {
        return Math.round(v * 1000.0) / 1000.0;
    }

    private static void trim(Deque<Long> times) {
        long cutoff = System.currentTimeMillis() - maxWindowMillis();
        while (!times.isEmpty() && times.peekFirst() < cutoff) {
            times.pollFirst();
        }
    }

    private static long maxWindowMillis() {
        int max = 0;
        for (int w : WINDOWS) {
            max = Math.max(max, w);
        }
        return max * 1000L;
    }

    private static int countWithin(Deque<Long> times, int windowSeconds) {
        long cutoff = System.currentTimeMillis() - windowSeconds * 1000L;
        int count = 0;
        for (Long t : times) {
            if (t >= cutoff) {
                count++;
            }
        }
        return count;
    }

    public Map<String, Object> snapshot() {
        synchronized (lock) {
            Map<String, Object> result = new LinkedHashMap<>();
            result.put("sent_total", sentTotal);
            result.put("recv_total", recvTotal);
            for (int window : WINDOWS) {
                result.put("sent_" + window + "s", countWithin(sentTimesMillis, window));
                result.put("recv_" + window + "s", countWithin(recvTimesMillis, window));
            }

            if (lastRecvMillis != null) {
                double secondsSince = (System.currentTimeMillis() - lastRecvMillis) / 1000.0;
                result.put("seconds_since_last_recv", Math.round(secondsSince * 10) / 10.0);
                result.put("last_recv_datetime", TIME_FMT.format(Instant.ofEpochMilli(lastRecvMillis)));
            } else {
                result.put("seconds_since_last_recv", null);
                result.put("last_recv_datetime", null);
            }

            if (yellowThresholdSeconds != null) {
                result.put("yellow_threshold_seconds", yellowThresholdSeconds);
            }

            if (!connections.isEmpty()) {
                result.put("connections", new LinkedHashMap<>(connections));
            }

            if (!gauges.isEmpty()) {
                result.put("gauges", new LinkedHashMap<>(gauges));
            }

            if (!latencies.isEmpty()) {
                Map<String, Object> latencyOut = new LinkedHashMap<>();
                for (Map.Entry<String, Deque<Double>> e : latencies.entrySet()) {
                    if (e.getValue().isEmpty()) {
                        continue;
                    }
                    List<Double> values = new ArrayList<>(e.getValue());
                    Collections.sort(values);
                    Map<String, Object> row = new LinkedHashMap<>();
                    row.put("count", values.size());
                    row.put("min_ms", round3(values.get(0)));
                    row.put("p50_ms", round3(percentile(values, 0.50)));
                    row.put("p95_ms", round3(percentile(values, 0.95)));
                    row.put("max_ms", round3(values.get(values.size() - 1)));
                    latencyOut.put(e.getKey(), row);
                }
                if (!latencyOut.isEmpty()) {
                    result.put("latency", latencyOut);
                }
            }

            return result;
        }
    }
}
